### DISCUSSION TITLE
xdotool windowactivate does not focus window

### DISCUSSION THREAD

**Discussion by rayvermey** [bakkeby/dwm-flexipatch#172] [Q&A]
Hi

hope you can help me.
If i do : xdotool search --name "Slack" windowactivate , the tag where Slack is does not get focus.
xdotool search --name "Slack" does give a PID though

What patches should i apply to make xdotool work?

Thank you.

Ray

**Comment by bakkeby** [comment]
The https://dwm.suckless.org/patches/focusonnetactive/ patch is what you would be looking for. It makes it so that the view changes to the tag the client is on when a \_NET\_ACTIVE\_WINDOW client message is received.

By default (without it) the client is only marked as urgent and the tag colours are inverted.

In dwm-flexipatch this can be enabled with the `FOCUSONNETACTIVE_PATCH` macro.

**Reply by rayvermey** [reply]
Thanks but also after enabling this patch the problem is still there.
Is it maybe not compatibe with another patch?
This are the enabled patches i have:

#define BAR_DWMBLOCKS_PATCH 1
#define BAR_FANCYBAR_PATCH 1
#define BAR_LAYOUTMENU_PATCH 1
#define BAR_LTSYMBOL_PATCH 1
#define BAR_TABGROUPS_PATCH 1
#define BAR_TAGGRID_PATCH 1
#define BAR_STATUS_PATCH 1
#define BAR_STATUSBUTTON_PATCH 1
#define BAR_SYSTRAY_PATCH 1
#define BAR_TAGS_PATCH 1
#define BAR_WINICON_PATCH 1
#define BAR_WINTITLE_PATCH 1
#define BAR_TITLE_LEFT_PAD_PATCH 1
#define BAR_ALTERNATIVE_TAGS_PATCH 1
#define BAR_COLOR_EMOJI_PATCH 1
#define BAR_EWMHTAGS_PATCH 1
#define BAR_HIDEVACANTTAGS_PATCH 1
#define BAR_PANGO_PATCH 1
#define BAR_STATUSCOLORS_PATCH 1
#define COOL_AUTOSTART_PATCH 1
#define CYCLELAYOUTS_PATCH 1
#define FAKEFULLSCREEN_CLIENT_PATCH 1
#define FLOATPOS_RESPECT_GAPS_PATCH 1
#define FOCUSADJACENTTAG_PATCH 1
#define FOCUSURGENT_PATCH 1
#define FOCUSONNETACTIVE_PATCH 1
#define FULLSCREEN_PATCH 1
#define IPC_PATCH 1
#define PERTAG_PATCH 1
#define PERTAG_VANITYGAPS_PATCH 1
#define REORGANIZETAGS_PATCH 1
#define SCRATCHPADS_PATCH 1
#define SCRATCHPAD_ALT_1_PATCH 1
#define SELFRESTART_PATCH 1
#define SHIFTVIEW_CLIENTS_PATCH 1
#define TAGALLMON_PATCH 1
#define TOGGLELAYOUT_PATCH 1
#define VANITYGAPS_PATCH 1
#define VANITYGAPS_MONOCLE_PATCH 1
#define VIEWONTAG_PATCH 1
#define BSTACK_LAYOUT 1
#define CENTEREDMASTER_LAYOUT 1
#define COLUMNS_LAYOUT 1
#define DECK_LAYOUT 1
#define FLEXTILE_DELUXE_LAYOUT 1
#define GRIDMODE_LAYOUT 1
#define HORIZGRID_LAYOUT 1
#define TILE_LAYOUT 1
#define MONOCLE_LAYOUT 1

Thanks

Ray

**Reply by bakkeby** [reply]
No the focusonnetactive patch should not be conflicting with anything else and it should be all that you need (for this feature anyway).

It does depend on a \_NET\_ACTIVE\_WINDOW client message being sent for the window though. Perhaps the way it is installed makes a difference, e.g. flatpak vs snap vs appimage vs packages.

If I run that search then I get four window IDs for Slack, so which one of those is chosen when you run `xdotool search --name "Slack" windowactivate` would also be relevant.

The first time I ran the command the slack window was not focused, but on subsequent calls it was so...

You can double check with other programs to confirm.

**Reply by rayvermey** [reply]
Can you show me your activated patches so i can compare?
Whatever i do, even firefox, st, slack no one focusses.
Ray

**Reply by bakkeby** [reply]
Besides the defaults:

```c
#define FOCUSONNETACTIVE_PATCH 1
```

What actually happens when you run your commands?

It is probably better to do a search using the **class** hint rather than the name as the latter will also match the terminal you are running the command from if the terminal name changes to that of the command being run.

```sh
$ xdotool search --name "Slack"
31457282
123731969
119537665
119537667
119537676 

$ xdotool getactivewindow
31457282 <---

$ xdotool search --class "Slack"
123731969
119537665
119537667
```

Also it is more common to use xdo for this instead of xdotool, e.g. this command will activate slack or start it if it is not running:

```sh
$ xdo activate -N Slack || slack
```
(the `-N` flag here means class)

Same with

```sh
$ xdo activate -N firefox || firefox
```
(just that the class hint for the browser for some reason tend to vary from distribution to distribution)

**Reply by rayvermey** [reply]
Hi, thanks for your help.
The xdotool search --name slack gives the PID number(s)
The xdotool search --name slack windowactivate just gives back prompt and nothing happens
xdo command same.
I am on a windowmanager called InstantWM (part of InstantOS) Arch based distro.
I only use the WM part since my system is plain Arch.
On instantWM xdotool works as expected.
On dwm-flexipatch it just doesnt.
The windowactivation just does not happen.
Could it be an ENVIRONMENT variable missing?
I mean what on earth could be the difference where xdotool does not work?
I have a 3 monitor setup btw...

Thanks again.

**Reply by bakkeby** [reply]
> The xdotool search --name slack gives the PID number(s)

No it doesn't. It gives WID numbers (window IDs) where the window name matches the string used as a search term.

> xdotool search --name slack windowactivate just gives back prompt and nothing happens

As stated earlier it could be that it selects the terminal you are running the command from and gives focus to that. Did you try with the class instead?

> On dwm-flexipatch it just doesnt.

There isn't anything special in dwm-flexipatch or the InstantWM build with regards to this, both just have the focusonnetactive patch applied.

Sanity check that you have recompiled and restarted dwm, beyond that I am not sure what to advice.

You would expect this to work with a bare dwm and that patch applied, which should be relatively easy to test.

> Could it be an ENVIRONMENT variable missing?

No there should be no environment variables involved, unless you are running the window manager through Xephyr of course.

**Reply by rayvermey** [reply]
I downloaded dwm-6.2.tar file
Unpacked
make (no errors)
sudo make install (no errirs)
.xinitrc "exec dwm"
And restarted
And now dwm will not start throwing XF86kbd errors at me.
Could that be an indication?
Cause other wm's work....
This problem is soooo strange.

**Reply by rayvermey** [reply]
Here a screenshot of what happens when i want to start dwm6.2
![unnamed](https://user-images.githubusercontent.com/42524768/127767824-c9688ab0-485e-4071-bf26-1c479b29257c.jpg)

**Reply by bakkeby** [reply]
A bare dwm does not use XF86 keys like XF86BrightnessAuto, so unless you copied in another config.h file then it would have to be related to something in your .xinitrc file.

NB: What the Kleenex is used for? :)

**Reply by rayvermey** [reply]
LOL am a vaper (electronic smoking) and those things leak sometimes.
OK
So my .xinitrc is empty.
My zshrc not...
Trying again with bare login session.
Then i must go, hope i catch you later today.
Thanks again!

**Reply by rayvermey** [reply]
Back home..
What i wonder is: why does "exec instantwm" works, but exec "dwm" give errors on startx...
dwm just compiles right out of the box.
So why would dwm not start???
Do you have any ideas maybe?

**Reply by rayvermey** [reply]
hmmmm i have the dwm-flexipatch working now!
I have patches.h and patches.def.h 
patches.h had not the FOCUSONNETACTIVE macro set to 1
patches.def.h did.
When i changed patches.h , compiled, then xdotool worked!
Still strange that fresh dwm does not start but hey, dwm-flexipatch is much better! :-)
Thanks for your help!

Ray

**Reply by bakkeby** [reply]
The `config.def.h` file as the name suggests is the default configuration file in dwm (and most suckless tools for that matter).

When you (first) compile dwm `config.def.h` is copied to `config.h` if the latter does not exist. `config.h` is then a local / personal configuration file that you can change as you want. This is often a source of confusion for new users as patches will apply to the default configuration and not your personal configuration file which again leads to compilation errors.

The `patches.def.h` file follows the same pattern in that the default `patches.def.h` file is copied to `patches.h` if it does not already exist.

**Reply by rayvermey** [reply]
OK so if i understand this correct: you leave the config.def.h alone, edit config.h for personal changes, then if there is a patch or an update, i must diff the config.def.h with my config.h and adjust my config.h accordingly to the patches or update?

**Reply by bakkeby** [reply]
Yes that is correct. Diffuse and meld are useful tools for this.

It is perfectly possible to hack it so that it always reads from `config.def.h` instead of `config.h` of course. It adds a risk of conflicts when it comes to updates, but this should for the most part be avoidable if you update using `git pull --rebase`.
